// apemanerror/shim/browser/buildin.js
/**
 * Buildin error creators
 * @module buildin
 */
'use strict';

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var apemanerror = require('./apemanerror');

module.exports = (0, _assign2.default)(exports, {},
// Basic errors
{
  valueMissingError: function valueMissingError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Value required';
    return apemanerror(msg, {
      name: 'ValueMissingError',
      status: 400
    });
  },
  valueInvalidError: function valueInvalidError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Value invalid';
    return apemanerror(msg, {
      name: 'valueInvalidError',
      status: 400
    });
  }
},
// Sign errors
{
  signupProveError: function signupProveError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Provision Failed';
    return apemanerror(msg, {
      name: 'SignupProveError',
      status: 400
    });
  },

  signinFailError: function signinFailError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Signin Failed';
    return apemanerror(msg, {
      name: 'SigninFail',
      status: 400
    });
  },

  signinRequiredError: function signinRequiredError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Signin required';
    return apemanerror(msg, {
      name: 'SigninRequiredError',
      status: 401
    });
  },

  signInvalidError: function signInvalidError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'User not registered for sign';
    return apemanerror(msg, {
      name: 'SignInvalidError',
      status: 403
    });
  }
},
// Verify errors
{
  verifyNotFoundError: function verifyNotFoundError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'User not found for the given email';
    return apemanerror(msg, {
      name: 'VerifyNotFoundError',
      status: 400
    });
  },

  verifyTokenWrongError: function verifyTokenWrongError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Token is wrong';
    return apemanerror(msg, {
      name: 'VerifyTokenWrongError',
      status: 400
    });
  },

  verifyTokenExpiredError: function verifyTokenExpiredError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Token is expired';
    return apemanerror(msg, {
      name: 'VerifyTokenExpiredError',
      status: 400
    });
  }
},
// Recover errors
{
  recoverNotFoundError: function recoverNotFoundError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'User not found for the given email';
    return apemanerror(msg, {
      name: 'RecoverNotFoundError',
      status: 400
    });
  },

  recoverTokenWrongError: function recoverTokenWrongError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Token is wrong';
    return apemanerror(msg, {
      name: 'RecoverTokenWrongError',
      status: 400
    });
  },

  recoverTokenExpiredError: function recoverTokenExpiredError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Token is expired';
    return apemanerror(msg, {
      name: 'RecoverTokenExpiredError',
      status: 400
    });
  }
},
// Auth errors
{
  notAllowedError: function notAllowedError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Not allowed here';
    return apemanerror(msg, {
      name: 'NotAllowedError',
      status: 403
    });
  }
},
// Resource errors
{
  notFoundError: function notFoundError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Not found';
    return apemanerror(msg, {
      name: 'NotFound',
      status: 404
    });
  }
},
// Authorize errors
{
  notAuthorizedError: function notAuthorizedError() {
    var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Not authroized';
    return apemanerror(msg, {
      name: 'NotAuthorized',
      status: 402
    });
  }
});
